File: android/src/main/java/com/photovideoeditor/photo/render/PhotoAdjustments.kt
package com.photovideoeditor.photo.render

/**
 * Non-destructive photo adjustment stack (Milestone 3). Every numeric field
 * has a documented range and a zero/no-op default so "reset" is always just
 * `PhotoAdjustments()`.
 */
data class PhotoAdjustments(
  val brightness: Float = 0f, // -100..100
  val contrast: Float = 0f, // -100..100
  val saturation: Float = 0f, // -100..100
  val exposure: Float = 0f, // -100..100
  val temperature: Float = 0f, // -100..100
  val blurRadius: Float = 0f, // 0..25 (px)
  val mirror: Boolean = false,
  val filterPreset: String? = null,
  val filterStrength: Float = 100f // 0..100, only meaningful when filterPreset is set
) {
  val isIdentity: Boolean get() = this == PhotoAdjustments()

  fun value(key: String): Float = when (key) {
    "brightness" -> brightness
    "contrast" -> contrast
    "saturation" -> saturation
    "exposure" -> exposure
    "temperature" -> temperature
    "blurRadius" -> blurRadius
    "filterStrength" -> filterStrength
    else -> 0f
  }

  fun with(key: String, value: Float): PhotoAdjustments = when (key) {
    "brightness" -> copy(brightness = value)
    "contrast" -> copy(contrast = value)
    "saturation" -> copy(saturation = value)
    "exposure" -> copy(exposure = value)
    "temperature" -> copy(temperature = value)
    "blurRadius" -> copy(blurRadius = value)
    "filterStrength" -> copy(filterStrength = value)
    else -> this
  }

  companion object {
    /** Adjustment keys that map to a continuous slider (excludes mirror/filterPreset, which are selectors). */
    val SLIDER_KEYS = listOf(
      "brightness", "contrast", "saturation", "exposure", "temperature", "blurRadius"
    )

    fun rangeFor(key: String): Pair<Float, Float> = when (key) {
      "blurRadius" -> 0f to 25f
      "filterStrength" -> 0f to 100f
      else -> -100f to 100f
    }
  }
}

File: ios/Photo/Rendering/PhotoAdjustments.swift
import Foundation

/// Non-destructive photo adjustment stack (Milestone 3). Every numeric field
/// has a documented range and a zero/no-op default so "reset" is always just
/// `PhotoAdjustments()`. Mirrors `PhotoAdjustments.kt` field-for-field so the
/// two platforms can share the same formulas in their renderers.
struct PhotoAdjustments: Equatable {
  var brightness: Double = 0 // -100..100
  var contrast: Double = 0 // -100..100
  var saturation: Double = 0 // -100..100
  var exposure: Double = 0 // -100..100
  var temperature: Double = 0 // -100..100
  var blurRadius: Double = 0 // 0..25 (px)
  var mirror: Bool = false
  var filterPreset: String?
  var filterStrength: Double = 100 // 0..100, only meaningful when filterPreset is set

  var isIdentity: Bool { self == PhotoAdjustments() }

  func value(_ key: String) -> Double {
    switch key {
    case "brightness": return brightness
    case "contrast": return contrast
    case "saturation": return saturation
    case "exposure": return exposure
    case "temperature": return temperature
    case "blurRadius": return blurRadius
    case "filterStrength": return filterStrength
    default: return 0
    }
  }

  mutating func setValue(_ key: String, _ newValue: Double) {
    switch key {
    case "brightness": brightness = newValue
    case "contrast": contrast = newValue
    case "saturation": saturation = newValue
    case "exposure": exposure = newValue
    case "temperature": temperature = newValue
    case "blurRadius": blurRadius = newValue
    case "filterStrength": filterStrength = newValue
    default: break
    }
  }

  static let sliderKeys = [
    "brightness", "contrast", "saturation", "exposure", "temperature", "blurRadius",
  ]

  static func range(for key: String) -> (Double, Double) {
    switch key {
    case "blurRadius": return (0, 25)
    case "filterStrength": return (0, 100)
    default: return (-100, 100)
    }
  }
}
